File: apps/desktop/src/modules/db/tools.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

/** The three sets of tools, one per SQL-or-document engine: each carries the pair its database
 *  needs. Only two of them are ever downloadable — PostgreSQL publishes no client-only archive. */
export type ToolSuite = "mysql" | "postgres" | "mongo";

/** Where a tool was found. `custom` is a path picked in Settings, `downloaded` a copy MixDB
 *  fetched for itself, and `system` something already installed on the machine. */
export type ToolSource = "custom" | "downloaded" | "system";

export interface ToolStatus {
  /** `mysqldump`, `mysql`, `mongodump` or `mongorestore`. */
  name: string;
  suite: ToolSuite;
  /** Where it is, or null when it is nowhere to be found. */
  path: string | null;
  source: ToolSource | null;
  /** Whether MixDB can fetch this tool for itself here. The same answer for every tool of a
   *  suite — MySQL publishes a plain archive for Windows only, so on macOS and Linux its tools
   *  have to come from the machine. */
  downloadable: boolean;
}

export function toolsStatus(): Promise<ToolStatus[]> {
  return invoke<ToolStatus[]>("dumptools_status");
}

/** Whether both of a suite's tools can be found — what the dump and restore buttons ask first. */
export function toolsReady(suite: ToolSuite): Promise<boolean> {
  return invoke<boolean>("dumptools_ready", { suite });
}

/** Whether there is anywhere to download this suite from on this platform — asked before a button
 *  offers to, since a suite with no archive can only be answered with an error. */
export function toolsDownloadable(suite: ToolSuite): Promise<boolean> {
  return invoke<boolean>("dumptools_downloadable", { suite });
}

/** The stages an install goes through, in the order they happen. */
export type ToolStage = "downloading" | "verifying" | "unpacking" | "installing";

/** How far an install has got. `total` is 0 when the server never said how big the archive is,
 *  which is the difference between a bar with a percentage and one that only shows movement. */
export interface ToolProgress {
  suite: ToolSuite;
  stage: ToolStage;
  done: number;
  total: number;
}

/** Which suites are being fetched, and which have just been. */
export interface ToolInstallState {
  /** Every suite being fetched right now, against how far it has got — a suite with no progress
   *  yet is in the map with `null`, which is what tells "starting" from "not running". */
  running: ReadonlyMap<ToolSuite, ToolProgress | null>;
  /** When each suite's install finished, for the line that says it did. */
  finished: ReadonlyMap<ToolSuite, number>;
}

/*
 * An install runs for minutes behind a modal dialog, which is to say behind the whole app: the
 * user will close Settings and get on with something else. So how far it has got is kept here,
 * out of the screen showing it, and the screen reads it back on the way in. Otherwise closing the
 * dialog would lose the download from sight and leave the button ready to start a second one.
 *
 * Keyed by suite rather than held one at a time: the two suites are downloaded to staging
 * directories of their own and unpacked into directories of their own, so fetching both at once is
 * two errands that never meet, and there is no reason to make the user wait out the first.
 */
let state: ToolInstallState = { running: new Map(), finished: new Map() };
const listeners = new Set<() => void>();
let listening: Promise<UnlistenFn> | null = null;

function setState(next: ToolInstallState) {
  state = next;
  for (const listener of listeners) listener();
}

/** The state as it stands with one suite's entries changed. `undefined` removes the entry. */
function withSuite(
  running?: [ToolSuite, ToolProgress | null | undefined],
  finished?: [ToolSuite, number | undefined],
): ToolInstallState {
  const nextRunning = new Map(state.running);
  const nextFinished = new Map(state.finished);
  if (running) {
    const [suite, progress] = running;
    if (progress === undefined) nextRunning.delete(suite);
    else nextRunning.set(suite, progress);
  }
  if (finished) {
    const [suite, at] = finished;
    if (at === undefined) nextFinished.delete(suite);
    else nextFinished.set(suite, at);
  }
  return { running: nextRunning, finished: nextFinished };
}

/** Registers the progress listener, once, and resolves when it is in place. */
function startListening(): Promise<UnlistenFn> {
  listening ??= listen<ToolProgress>("tools://progress", ({ payload }) => {
    // An install that has already ended has nothing left to report; ignore anything still in
    // flight behind it rather than reviving the bar.
    if (!state.running.has(payload.suite)) return;
    setState(withSuite([payload.suite, payload]));
  });
  return listening;
}

/** How the install is going, for `useSyncExternalStore`. The same object until something changes,
 *  which is what stops it from re-rendering on every read. */
export function toolInstallState(): ToolInstallState {
  return state;
}

export function subscribeToolInstall(listener: () => void): () => void {
  listeners.add(listener);
  return () => {
    listeners.delete(listener);
  };
}

/** Downloads a suite from its vendor. Tens of megabytes at best, so it reports its way through
 *  {@link toolInstallState} — and resolving is what says it finished. */
export async function toolsInstall(suite: ToolSuite): Promise<void> {
  // Before the download starts, so the first bytes are not counted against an unregistered
  // listener.
  await startListening();
  setState(withSuite([suite, null], [suite, undefined]));
  try {
    await invoke<void>("dumptools_install", { suite });
    setState(withSuite([suite, undefined], [suite, Date.now()]));
  } catch (e) {
    setState(withSuite([suite, undefined]));
    throw e;
  }
}

/** Deletes MixDB's own copy. Tools found on the machine itself are left alone. */
export function toolsUninstall(suite: ToolSuite): Promise<void> {
  // "Downloaded and ready to use" stops being true the moment the user asks for it to go, so the
  // line saying so goes with the click rather than waiting out the rest of its few seconds.
  if (state.finished.has(suite)) setState(withSuite(undefined, [suite, undefined]));
  return invoke<void>("dumptools_uninstall", { suite });
}

/** Points one tool at a copy chosen by hand, or forgets that choice when given null. */
export function toolsSetPath(tool: string, path: string | null): Promise<void> {
  return invoke<void>("dumptools_set_path", { tool, path });
}

File: apps/desktop/src-tauri/src/modules/db/commands/tools.rs
//! The dump and restore tools: what MixDB can find, download and be pointed at.
//!
//! Commands of this module rather than of the app: a suite is one engine's pair of
//! programs (`mysqldump` and `mysql`, `pg_dump` and `psql`, `mongodump` and
//! `mongorestore`), and nothing outside the database module has a use for them.
//!
//! Tên lệnh mang tiền tố `dumptools_` chứ không phải `tools_`: module Tools có lệnh riêng của nó,
//! và `generate_handler!` là một danh sách phẳng nên hai họ lệnh khác chủ mà trùng tiền tố sẽ nằm
//! lẫn vào nhau mà không có gì cảnh báo.

use crate::error::AppError;
use tauri::{AppHandle, Emitter};
use crate::modules::db::drivers::tools;
use super::{in_background, tools_dir, TOOLS_PROGRESS_EVENT};

/// Every dump tool and where it stands: a path the user chose, a copy MixDB downloaded, something
/// already on the machine, or nothing at all.
///
/// Off the runtime, like everything else here: finding a tool walks `PATH` stat-ing candidates,
/// and one entry on an unreachable network share is enough to hang whichever thread does it.
#[tauri::command]
pub async fn dumptools_status(app: AppHandle) -> Result<Vec<tools::ToolStatus>, AppError> {
    let dir = tools_dir(&app)?;
    in_background(move || Ok(tools::status(&dir))).await
}

/// Whether a suite is usable at all — what the dump and restore buttons check before running.
#[tauri::command]
pub async fn dumptools_ready(app: AppHandle, suite: String) -> Result<bool, AppError> {
    let suite = tools::Suite::parse(&suite)?;
    let dir = tools_dir(&app)?;
    in_background(move || Ok(tools::installed(suite, &dir))).await
}

/// Whether MixDB can fetch this suite for itself on this platform — MySQL publishes a plain
/// archive for Windows only, so everywhere else its tools have to come from the machine.
#[tauri::command]
pub async fn dumptools_downloadable(suite: String) -> Result<bool, AppError> {
    Ok(tools::downloadable(tools::Suite::parse(&suite)?))
}

/// Points a tool at a copy the user picked themselves, or forgets that choice when given no path.
#[tauri::command]
pub async fn dumptools_set_path(
    app: AppHandle,
    tool: String,
    path: Option<String>,
) -> Result<(), AppError> {
    let tool = tools::Tool::parse(&tool)?;
    let dir = tools_dir(&app)?;
    in_background(move || tools::set_path(tool, path.as_deref(), &dir)).await
}

/// Deletes the copy MixDB downloaded. What was already on the machine is left where it is.
#[tauri::command]
pub async fn dumptools_uninstall(app: AppHandle, suite: String) -> Result<(), AppError> {
    let suite = tools::Suite::parse(&suite)?;
    let dir = tools_dir(&app)?;
    in_background(move || tools::uninstall(suite, &dir)).await
}

/// Downloads one suite of tools.
///
/// Minutes long on an ordinary connection, so it reports as it goes: every stage, and a running
/// byte count while the archive comes down, on `tools://progress`. The command returning is what
/// says it is finished — the events only say how far along it is.
#[tauri::command]
pub async fn dumptools_install(app: AppHandle, suite: String) -> Result<(), AppError> {
    let suite = tools::Suite::parse(&suite)?;
    let dir = tools_dir(&app)?;
    let reporter = app.clone();
    in_background(move || {
        tools::install(suite, &dir, &|progress| {
            // A dropped progress event is not worth failing an install over: the next one is a
            // quarter of a second away, and the last word comes from the command's own result.
            let _ = reporter.emit(TOOLS_PROGRESS_EVENT, progress);
        })
    })
    .await
}
